fix: edits to activity 3 and supporting lesson (#21)
pyOpenSci/lessons: clean-modular-code/activity-3/clean-code-activity-3.md

Writing robust code that handles unexpected values will make your code run smoothly and fail gracefully. This type of code, which combines functions (or classes) and checks within the functions that handle messy data, will make your code easier to maintain.

Things like helpful error messages and fast failing will also improve the experience for someone else using your code - OR your future self. 
 
:::{tip}
Using functions, classes, and methods (functions within a class) is a great first step in handling messy data. A function or method provides a modular unit you can test outside of the workflow for the edge cases you may encounter. Also, because a function is a modular unit, you can add elements to handle unexpected processing features as you build your workflow.

Once you have these functions and methods, you can add checks using conditional statements and try/except blocks that anticipate edge cases and errors that you may encounter when processing your data. 
:::

## Manage the unexpected 

In this activity, you will apply the following strategies to make your code more robust, maintainable & usable:

> * **[Fail fast with useful error messages](fail-fast)**: Failing fast is a software engineering term that means allowing your 
>   code to stop early when something goes wrong, ensuring that errors are caught 
>   and communicated clearly. This helps the user quickly understand what went 
>   wrong and where.

Reviewer: ```suggestion
* **[Fail fast with useful error messages](fail-fast)**: Failing fast is a software engineering term that means allowing your 
  code to stop when something goes wrong, ensuring that errors are caught 
  and communicated promptly. This helps the user quickly understand the error, what went 
  wrong, and where.
```

> * Use **[conditional statements](../checks-conditionals/python-conditionals)** 
>   to check for specific conditions before executing certain parts. This allows you to create specific workflows that your code followed based on specific conditions.

Reviewer: ```suggestion
* Use **[conditional statements](../checks-conditionals/python-conditionals)** 
  to logically check for specific conditions before executing code. This allows you to create different pathways for code to execute based on specific conditions.
```

* **[Try/except blocks](../checks-conditionals/python-function-checks)** allow 
  you to handle potential errors by attempting an operation and catching any 
  exceptions if they occur, providing useful feedback. In some cases you may want the program to end on an error. In other cases, you may want to handle it in a specific way. 

When you can, try to use a [Pythonic approach](pythonic-checks) to catch errors early; this means asking for forgiveness later (often using try/except blocks) vs. using conditional statements to check an object's state or type.

```{code-cell} ipython3
---
editable: true
slideshow:
  slide_type: ''
---

> # This works but is less Pythonic as it's a look before you leap approach

Reviewer: ```suggestion
# This works but is less Pythonic as it's a "look before you leap" approach since it does an extra check before returning the title
```

    return title
```

> The function below raises an error with a custom error message. but you can still see the 

Reviewer: ```suggestion
The function below raises an error with a custom error message. but you can still see the error's detailed traceback
```

# This is the preferred way to catch an error 
def clean_title(title):
    """
    Attempts to return the first character of the title.

>     Raises the same error with a friendly message if the input is invalid.

Reviewer: ```suggestion
    Raises the same error with a friendly, custom error message if the input is invalid.
```

    """
    try:
        return title[0]
    except IndexError as e:
        raise IndexError(f"Oops! You provided a title in an unexpected format. "
                         f"I expected the title to be provided in a list and you provided "
                         f"a {type(title)}.") from e

# Example usage:
title = ""
print(clean_title(title))  # This will raise an IndexError with the friendly message
```

```{code-cell} ipython3
# This is the preferred way to catch an error 
def clean_title(title):
    """
    Attempts to return the first character of the title.

>     Raises the same error with a friendly message if the input is invalid.

Reviewer: ```suggestion

    Raises the same error with a friendly message if the input is invalid.
```

    """
    try:
        return title[0]
    except IndexError as e:
        raise IndexError(f"Oops! You provided a title in an unexpected format. I expected the title to be provided in a list and you provided a {type(title)}.") 

# Example usage:
title = ""
print(clean_title(title))  # This will raise an IndexError with the friendly message
```

> If you wish, you can shorten the amount of information returned in the exception by adding `from None` to your exception. This will look nicer to a user but you lose some information in the exception feedback. 

Reviewer: ```suggestion
If you wish, you can shorten the amount of information returned in the error by adding `from None` when you raise the error. This will look nicer to a user, but you lose some detail in the error traceback. 
```

```{code-cell} ipython3
# This is the preferred way to catch an error 
def clean_title(title):
    """
    Attempts to return the first character of the title.
    Raises the same error with a friendly message if the input is invalid.
    """
    try:
        return title[0]
    except IndexError as e:
        raise IndexError(f"Oops! You provided a title in an unexpected format. "
                         f"I expected the title to be provided in a list and you provided "
                         f"a {type(title)}.") from None 

# Example usage:
title = ""
print(clean_title(title))  # This will raise an IndexError with the friendly message
```

+++ {"editable": true, "slideshow": {"slide_type": ""}, "tags": ["hide-output", "hide-cell"]}

> In this activity, you will be working with Pandas dataframes. You may find the .apply function to be particularly useful. 

Reviewer: ```suggestion
In this activity, you will be working with Pandas dataframes. You may find the `.apply` function to be particularly useful. 
```

+++ {"editable": true, "slideshow": {"slide_type": ""}}

> ## What i don't like

Reviewer: ```suggestion
## What I don't like
```
